tests/test_accounts_service.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/python3
+""" Tests for accounts service """
+
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the Free
+# Software Foundation; either version 3 of the License, or (at your option) any
+# later version.  See http://www.gnu.org/copyleft/lgpl.html for the full text
+# of the license.
+
+__author__ = 'Marco Trevisan'
+__copyright__ = '(c) 2021 Canonical Ltd.'
+
+import subprocess
+import sys
+import unittest
+
+import dbusmock
+
+try:
+    import gi  # type: ignore
+    gi.require_version('AccountsService', '1.0')
+    from gi.repository import AccountsService, GLib
+    have_accounts_service = True
+except (ImportError, ValueError):
+    have_accounts_service = False
+
+
+@unittest.skipUnless(have_accounts_service,
+                     'AccountsService gi introspection not available')
+class TestAccountsService(dbusmock.DBusTestCase):
+    '''Test mocking AccountsService'''
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.start_system_bus()
+        cls.dbus_con = cls.get_dbus(True)
+        cls.ctx = GLib.main_context_default()
+
+    def setUp(self):
+        super().setUp()
+        (self.p_mock, self.p_obj) = self.spawn_server_template(
+            'accounts_service', {}, stdout=subprocess.PIPE)
+        self.p_manager = AccountsService.UserManager.get_default()
+        while not self.p_manager.props.is_loaded:
+            self.ctx.iteration(True)
+        self.assertFalse(self.p_manager.no_service())
+        self.assertTrue(self.p_manager.props.is_loaded)
+
+    def tearDown(self):
+        for user in self.p_manager.list_users():
+            self.p_manager.delete_user(user, False)
+
+        while self.p_manager.list_users():
+            self.ctx.iteration(True)
+
+        if self.p_mock:
+            self.p_mock.stdout.close()
+            self.p_mock.terminate()
+            self.p_mock.wait()
+
+        while self.p_manager.props.is_loaded:
+            self.ctx.iteration(True)
+
+        self.assertFalse(self.p_manager.props.is_loaded)
+        del self.p_manager
+        super().tearDown()
+
+    def wait_changed(self, user):
+        changed = False
+
+        def on_changed(u):
+            nonlocal changed
+            changed = u is user
+        conn_id = user.connect('changed', on_changed)
+        while not changed:
+            self.ctx.iteration(True)
+        user.disconnect(conn_id)
+
+    def test_empty(self):
+        self.assertTrue(self.p_manager.props.is_loaded)
+        self.assertFalse(self.p_manager.list_users())
+        self.assertFalse(self.p_manager.props.has_multiple_users)
```

Comment:
check has_no_users as well here?

Reply:
That's not exported via Gir, so in fact I was thinking to test this via the actual dbus api instead. I forgot to do it though.

Reply:
No, that's fine --  just check the official lib API.